### PR TITLE
BUGFIX: Ensure users content stream is never left closed after publication

### PR DESCRIPTION
**Pull request by mhsdesign**
Followup to publishing version 3: https://github.com/neos/neos-development-collection/pull/5301

There are 3 cases currently that might lead to a closed content stream if $something went wrong during publication.

### Case 1: Ensure all events are published BEFORE catchup

Otherwise, due to failures in projection or catchup-hooks the process would be immediately interrupted leaving a broken state.

For example a faulty redirect handler hook - that just listens to live events - would be called during publishing. That means the remaining part to publish is already commited and we know we still have work to do to fork the new user content stream and apply the remaining. But the catchup hook would interrupt immediately when the events were catchup'd live. We would be left with a CLOSED user content stream that contains the "same" events that went live during the rebase. Reopening would not help at that point. This is why we must ensure that all events are published BEFORE we do the first catchup.

Further implications:
- running catchup only once should be more performant
- we cannot refetch the current content stream version for what where previously "subcommans" (`forkContentStream`) but we must pass $expectedVersions around from the outside
- we should not run constraint checks after the first `yield` as that would still operate on the old state. Thus all checks are combined above
- closing the content stream will not be catch up'd directly meaning that another process will not get an error because the content stream is not writeable but because due to the version in the database being behind it will get a concurrency exception

### Case 2: Reopen content stream if base workspace was written to during publication

... and a ConcurrencyException is thrown

Introduces a `WorkspacePublicationDuringWritingTest` parallel test (with own cr) to assert that behaviour.

### Case 3: Close content stream a bit later instead of having to reopen in unexpected errors

Previously an error in `extractFromEventStream` because the payload was not correct and yet has to be migrated would lead to a closed content stream which is of course persisted even after fixing the events via migration.

This is still save to do, as the `closeContentStream` will commit the close on the FIRSTly fetched expected version.

Same guarantees, different error behaviour in rare cases.

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php [hunk 1]
```diff
@@ -29,7 +30,10 @@ public function __construct(
     }
 
     /**
-     * @return EventsToPublish|\Generator<int, EventsToPublish>
+     * The handler only calculate which events they want to have published,
+     * but do not do the publishing themselves
+     *
+     * @return EventsToPublish|YieldedEventsToPublish
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php]: Is that really easier to read than
```suggestion
     * @return EventsToPublish|\Generator<EventsToPublish>
```
?

File: Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlerInterface.php [hunk 2]
```diff
@@ -23,7 +25,7 @@ public function canHandle(CommandInterface $command): bool;
      * For the case of the workspace command handler that need to publish to many streams and "close" the content-stream directly,
      * it's allowed to yield the events to interact with the control flow of event publishing.
      *
-     * @return EventsToPublish|\Generator<int, EventsToPublish>
+     * @return EventsToPublish|YieldedEventsToPublish
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/CommandHandler/CommandHandlerInterface.php]: same here, I think the custom type does more harm than it helps to be honest.
(but not a big issue of course)

File: Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php [hunk 3]
```diff
@@ -31,15 +32,22 @@ public function publishEvents(ContentRepository $contentRepository, EventsToPubl
...
             return;
         }
+        $this->publishWithoutCatchup($eventsToPublish);
+        $contentRepository->catchUpProjections();
+    }
+
+    /**
+     * @throws ConcurrencyException in case the expectedVersion does not match
+     */
+    public function publishWithoutCatchup(EventsToPublish $eventsToPublish): CommitResult
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php:41]: This makes the `EventPersister` even more ugly than it already is :)
But I am confident that we'll be able to simplify the whole thing with #5321

File: Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php [hunk 4]
```diff
@@ -67,29 +37,26 @@ private function closeContentStream(
...
+            ExpectedVersion::fromVersion($contentStreamVersion)
         );
     }
 
     /**
      * @param ContentStreamId $contentStreamId The id of the content stream to reopen
      * @phpstan-pure this method is pure, to persist the events they must be handled outside
      */
-    private function reopenContentStream(
+    private function reopenContentStreamWithoutConstraints(
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php]: I had suggested to rename this method but only if there's another version that _does_ constraint checks.
In this case I would just keep the previous name (and the new inline comment) since it's an implementation detail.
If you want to keep the new name it should be "...WithoutConstraintChecks"

File: Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php [hunk 5]
```diff
@@ -133,25 +91,19 @@ private function forkContentStream(
      * @param ContentStreamId $contentStreamId The id of the content stream to remove
      * @phpstan-pure this method is pure, to persist the events they must be handled outside
      */
-    private function removeContentStream(
+    private function removeContentStreamWithoutConstraints(
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/ContentStreamHandling.php]: Same here: I would only add the suffix if there's a version _with_ constraint checks

**Review by bwaidelich — COMMENTED** [review]
I think it's a great start and the LoC balance is negative even:
image

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 6]
```diff
@@ -94,19 +94,35 @@ public function __construct(
...
+                return;
+            }
+            $this->eventPersister->publishWithoutCatchup($eventsToPublish);
+            $this->catchupProjections();
+            return;
+        }
+
+        // control-flow aware command handling via generator
+        try {
+            $yieldedEventsToPublish = $toPublish->current();
```
Comment (kitsunet) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: The question is why would we then even need to yield, because we might as well return one EventsToPublish then, if there is no catch up in between?! Also means e.g. pre-emptive closing of CS is pointless as other processes constraint checks will never see it when it would be useful to have.

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: > pre-emptive closing of CS is pointless as other processes constraint checks will never see it when it would be useful to have.

we thought about that:

> > closing the content stream will not be catch up'd directly meaning that another process will not get an error because the content stream is not writeable but because due to the version in the database being behind it will get a concurrency exception


thats why we yield. Further - still todo - we need to reopen the user workspace for the unlikely case that the matching events could not be published, as we dont lock or close the base workspace ... maybe we should though ...

File: Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/CommandHandler/CommandBus.php]: adjusted :) (without complaining because i did all the argumentation myself in my head ... over dayyyyys :D)

File: .composer.json [hunk 7]
```diff
@@ -25,8 +25,9 @@
       "../../bin/phpunit --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/UnitTests.xml Neos.ContentRepository.Core/Tests/Unit",
       "../../bin/phpunit --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/UnitTests.xml Neos.ContentRepositoryRegistry/Tests/Unit"
     ],
+    "test:paratest-cli": "../../bin/paratest --debug -v --functional --processes 2 --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/FunctionalTests.xml",
     "test:parallel": [
-      "FLOW_CONTEXT=Testing/Behat ../../bin/paratest --debug -v --functional --group parallel --processes 2 --colors --stop-on-failure -c ../../Build/BuildEssentials/PhpUnit/FunctionalTests.xml Neos.ContentRepository.BehavioralTests/Tests/Functional/Feature/WorkspacePublication/WorkspaceWritingDuringPublication.php"
+      "for f in Neos.ContentRepository.BehavioralTests/Tests/Parallel/**/*Test.php; do composer test:paratest-cli $f; done"
```
Comment (bwaidelich) [inline .composer.json:30]: (how) is that related and what does it do?

File: Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php [hunk 8]
```diff
@@ -52,14 +52,15 @@ final protected function awaitFile(string $filename): void
...
-        while (!flock($resource, LOCK_SH)) {
-            usleep(10000);
+        while (!is_file($filename)) {
+            usleep(1000);
             $waiting++;
-            if ($waiting > $maximumCycles) {
-                throw new \Exception('timeout while waiting on shared lock');
+            clearstatcache(true, $filename);
+            if ($waiting > 60000) {
+                throw new \Exception('timeout while waiting on file ' . $filename);
```
Comment (bwaidelich) [inline Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php:63]: Should we maybe extract those changes to the infra/test setup to a separate pr? ( mostly to allow to get a useful log via git blame)

File: Neos.Neos/Classes/Domain/Model/UserId.php [hunk 9]
```diff
@@ -15,7 +15,7 @@ public function __construct(
         public string $value
     ) {
         if (!preg_match('/^([a-z0-9\-]{1,40})$/', $value)) {
-            throw new \InvalidArgumentException(sprintf('Invalid user id "%s" (a user id must only contain lowercase characters, numbers and the "-" sign).', 1718293224));
+            throw new \InvalidArgumentException(sprintf('Invalid user id "%s" (a user id must only contain lowercase characters, numbers and the "-" sign).', $this->value), 1718293224);
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Domain/Model/UserId.php:18]: Oops. Good catch

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 10]
```diff
@@ -94,19 +95,48 @@ public function __construct(
...
-                $eventsToPublish = $this->enrichEventsToPublishWithMetadata($eventsToPublish);
-                $this->eventPersister->publishEvents($this, $eventsToPublish);
+        $toPublish = $this->commandBus->handle($command);
+
+        if ($toPublish instanceof EventsToPublish) {
+            // simple case
+            $eventsToPublish = $this->enrichEventsToPublishWithMetadata($toPublish);
+            if ($eventsToPublish->events->isEmpty()) {
+                return;
+            }
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: This check could be moved before the enrichment I suppose
 But it's not important and it will hopefully be removed with my other pr anyways

File: Neos.ContentRepository.Core/Classes/ContentRepository.php [hunk 11]
```diff
@@ -94,19 +95,48 @@ public function __construct(
...
+            $this->eventPersister->publishWithoutKetchup($eventsToPublish);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/ContentRepository.php]: Did you really call it Ketchup to troll us? 🙃

File: Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php [hunk 12]
```diff
@@ -24,22 +25,31 @@ public function __construct(
...
         }
+        $this->publishWithoutKetchup($eventsToPublish);
+        $contentRepository->catchUpProjections();
+    }
+
+    /**
+     * TODO Will be refactored via https://github.com/neos/neos-development-collection/pull/5321
+     * @throws ConcurrencyException in case the expectedVersion does not match
+     */
+    public function publishWithoutKetchup(EventsToPublish $eventsToPublish): CommitResult
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php]: 🍅

File: Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php [hunk 13]
```diff
@@ -81,14 +81,9 @@ protected function requireContentStream(
         CommandHandlingDependencies $commandHandlingDependencies
     ): ContentStreamId {
         $contentStreamId = $commandHandlingDependencies->getContentGraph($workspaceName)->getContentStreamId();
-        if (!$commandHandlingDependencies->contentStreamExists($contentStreamId)) {
-            throw new ContentStreamDoesNotExistYet(
-                'Content stream for "' . $workspaceName->value . '" does not exist yet.',
-                1521386692
-            );
-        }
+        $closedState = $commandHandlingDependencies->isContentStreamClosed($contentStreamId);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php]: Nitpick: rename to `$contentStreamIsClosed`

**Review by bwaidelich — COMMENTED** [review]
Some, probably semi-helpful, comments. Looks great otherwise

File: .composer.json (hunk 7 above)
Comment (mhsdesign) [reply .composer.json:30]: another paratest so we need to execute all not only one. that was more tricky than i wanted it to be:

see

https://github.com/neos/neos-development-collection/pull/5342/commits/18072f695823924fae05c2455a8b356fab8a96a2

and my question here:

https://github.com/paratestphp/paratest/discussions/905

File: Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php (hunk 8 above)
Comment (mhsdesign) [reply Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php:63]: nein :D. We already have way to many prs and discussions flying around. 
I did fix the previous para test with this dedicated pr but imo its super fine to push a few minor fixes on it :)
https://github.com/neos/neos-development-collection/pull/5344

and i used dedicated commits ;)

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: Maybe for jon :D 
It will be removed either way with the subscription engine:

> TODO Will be refactored via https://github.com/neos/neos-development-collection/pull/5321

:D

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 10 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: adjusted :)

File: Neos.ContentRepository.Core/Classes/ContentRepository.php (hunk 11 above)
Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/ContentRepository.php]: adjusted as requested :) https://github.com/neos/neos-development-collection/pull/5342/commits/e12c6414aa5a89bc5835f002ef69abab4b12d76e

File: Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php (hunk 8 above)
Comment (bwaidelich) [reply Neos.ContentRepository.BehavioralTests/Tests/Parallel/AbstractParallelTestCase.php:63]: "nein" is not really an appropriate response to a suggestion in a code review 😆 
But I get your point, and thanks for the dedicated commits

**Review by bwaidelich — APPROVED** [review]
+1 by reading and running the tests

**Review by dlubitz — COMMENTED** [review]
Look good by reading. But as a disclaimer ... I'm not that deep into this topic to fully understand. 🤷‍♂️

File: Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php (hunk 12 above)
Comment (kitsunet) [reply Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php]: refactored or not, that other PR is still Draft, and while I have full trust in bwaidelich, I don't like to bet on anything here, we should release, maybe it must be without subscription engine 🤷 who knows, but I don't really awnt ot releaes with ketchup. So please rename :)

Comment (dlubitz) [reply Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php]: Thanks for pointing out how careful I did my review 🙈

Comment (dlubitz) [reply Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php]: I guess I had 🍅 🍅 on my eyes.

Comment (mhsdesign) [reply Neos.ContentRepository.Core/Classes/EventStore/EventPersister.php]: i renamed it again to the proper naming in https://github.com/neos/neos-development-collection/pull/5342/commits/dfd457360a13aee60ae9c22d9a19c5c4b5683dcb :D I had it corrected but was tempted as we had this fun sprint today and what would be a sprint day without Schabernack :D

**Review by kitsunet — APPROVED** [review]
